src/cunumeric/matrix/diag.cc

```diff
@@ -22,30 +22,55 @@ namespace cunumeric {
 using namespace Legion;
 using namespace legate;
 
-template <LegateTypeCode CODE>
-struct DiagImplBody<VariantKind::CPU, CODE> {
+template <LegateTypeCode CODE, int DIM>
+struct DiagImplBody<VariantKind::CPU, CODE, DIM, true> {
   using VAL = legate_type_of<CODE>;
 
-  void operator()(const AccessorWO<VAL, 2>& out,
-                  const AccessorRO<VAL, 2>& in,
-                  const Point<2>& start,
-                  size_t distance) const
+  void operator()(const AccessorRD<SumReduction<VAL>, true, DIM>& out,
+                  const AccessorRO<VAL, DIM>& in,
+                  const coord_t& start,
+                  Pitches<DIM - 1>& m_pitches,
+                  Rect<DIM>& m_shape,
+                  const size_t naxes,
+                  const coord_t distance) const
   {
-    for (coord_t idx = 0; idx < distance; ++idx) {
-      Point<2> p(start[0] + idx, start[1] + idx);
-      out[p] = in[p];
+    size_t skip_size = 1;
+
+    for (int i = 0; i < naxes; i++) {
+      auto diff = 1 + m_shape.hi[DIM - i - 1] - m_shape.lo[DIM - i - 1];
+      if (diff != 0) skip_size *= diff;
```

Comment:
Can this ever be 0?

Reply:
diff here is a size of a partition in a particular dimension (i). It can be 0, but we don't want to multiply skip_size by 0, instead we just skip this dimension